### PR TITLE
refactor(core): guard against `importProvidersFrom` in components

### PR DESCRIPTION
**Pull request by alxhub**
`importProvidersFrom` provides a bridge from the world of NgModule-based DI
configuration to the new, "standalone" world of direct providers and
environment injectors. Early user feedback suggested some confusion around
where this function was supposed to be used, particularly around importing
NgModule-based providers into standalone component `providers` arrays, which
is not the intended use. This confusion is exacerbated by the fact that due
to the unified `Provider` type, this kind of misconfiguration was happily
accepted by the type system.

This commit changes the return type of `importProvidersFrom` to wrap the
returned providers in an opaque type that prevents them from being used in
component provider contexts. This, together with stronger documentation
around the purpose and functionality of `importProvidersFrom`, should
address some of the above confusion.

### CODE REVIEW

File: packages/core/src/di/interface/provider.ts [hunk 1]
```diff
@@ -351,3 +351,20 @@ export interface ModuleWithProviders<T> {
...
   providers?: Provider[];
 }
+
+/**
+ * Providers that were imported from NgModules via the `importProvidersFrom` function.
+ *
+ * These providers are meant for use in an application injector (or other environment injectors) and
+ * should not be used in component injectors.
+ *
+ * This type cannot be directly implemented. It's returned from the `importProvidersFrom` function
```
Comment (pkozlowski-opensource) [inline packages/core/src/di/interface/provider.ts:361]: nit: "cannot" or "is not meant to" / "should not"? I mean, is there anything, technically speaking, that prevents people from implementing this interface?

**Review by pkozlowski-opensource — CHANGES_REQUESTED** [review]
LGTM overall but we need a test + runtime checks when someone "outsmarts" the type system.

**Review by pkozlowski-opensource — DISMISSED** [review]
Awesome, thnx for the additional check on the node injector side + test @alxhub !

LGTM

File: packages/core/src/di/interface/provider.ts (hunk 1 above)
Comment (alxhub) [reply packages/core/src/di/interface/provider.ts:361]: As far as a user is concerned, there's no way you can produce a value of this type. Yes, you could cast, or write the private prefixed property yourself, but this is documenting the restriction as it applies to a "law-abiding" user.

**Review by atscott — APPROVED** [review]
reviewed-for: public-api

**Review by atscott — DISMISSED** [review]
reviewed-for: fw-router

**Review by AndrewKushnir — DISMISSED** [review]
Reviewed-for: public-api

**Review by dylhunn — DISMISSED** [review]
reviewed-for: public-api

**Comment by dylhunn** [comment]
@alxhub Mind doing a quick rebase?

**Comment by AndrewKushnir** [comment]
@dylhunn @alxhub JFYI I've resolved merge conflicts and pushed a rebase.

**Comment by AndrewKushnir** [comment]
Merge-assistance: pushing a rebase caused the approvals to reset. No actual changes to the core files were made during the rebase: the conflict was in a spec file (in the file imports section).

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit d322052db3c66904010e03074b5c7b611e511b2d.

**Pull request closed without merging**